cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java

```diff
@@ -1011,21 +1012,31 @@ private QueryBuilder createMainQuery(SearchRequest request) {
     // all terms must occur in the text field as selected by the scope in the search request.
     String textField = request.getScope() == null ? TEXT_FIELD : request.getScope().name().toLowerCase();
 
-    // split the query into its terms and iterate over all terms
-    Iterable<String> terms = Splitter.on(SPACE_SEPARATOR_PATTERN)
-      .omitEmptyStrings().trimResults().split(request.getQuery());
-    List<QueryBuilder> termQueries = new ArrayList<>();
-    for (String term : terms) {
-      termQueries.add(createTermQuery(term, textField, request));
+    // split the query into its parsed terms and create corresponding QueryBuilders for each
+    ArrayList<ImmutablePair<QueryBuilder, QueryTerm>> termQueries = new ArrayList<>();
```

Comment:
Instead of storing all terms in an intermediate List, I would suggest the following logic, which I think is cleaner.

```java
List<QueryTerm> queryTerms = QueryParser.parse(request.getQuery());
if (queryTerms.isEmpty()) {
  return QueryBuilders.matchAllQuery();
}
if (queryTerms.size() == 1) {
  return createTermQuery(queryTerms.get(0));
}

BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
for (QueryTerm queryTerm : queryTerms) {
  QueryBuilder builder = createTermQuery(queryTerm);
  if (queryTerm.getQualifier() == Qualifier.REQUIRED) {
    boolQuery.must(builder);
  } else {
    boolQuery.should(builder);
  }
}

```

Reply:
I agree. Added to latest commit.